tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java

```diff
@@ -0,0 +1,259 @@
+            } catch (InterruptedException e) {
+                log.info("{}: StatusUpdater is interrupted.", id);
+            } catch (Exception e) {
+                WorkerUtils.abort(log, "Prepare", e, doneFuture);
+            } finally {
+                if (process != null && process.isAlive()) {
+                    process.destroy();
+                }
+            }
+        }
+    }
+
+    public class ControlCommand {
+        private final String action;
+        private final ExternalCommandSpec spec;
+        @JsonCreator
+        ControlCommand(@JsonProperty("action") String action,
+                       @JsonProperty("spec") ExternalCommandSpec spec) {
+            this.action = action;
+            this.spec = spec;
+        }
+        @JsonProperty
+        public String action() {
+            return action;
+        }
+        @JsonProperty
+        public ExternalCommandSpec spec() {
+            return spec;
+        }
+
+        public void issue() {
+            String command = JsonUtil.toJsonString(this);
+            log.info("{}: Send out command: {} ", id, command);
+            controlChannel.println(command);
+        }
+    }
+
+    public class ErrorUpdater implements Runnable {
+        @Override
+        public void run() {
+            try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8))) {
+                String line;
+                while ((line = br.readLine()) != null) {
+                    log.error("{}: (stderr of the process):{}", id, line);
+                    errMsg = line;
+                }
+            } catch (IOException ioe) {
+                log.info("{}: Stderr of the process is closed.", id);
+            } catch (Exception e) {
+                WorkerUtils.abort(log, "ErrorUpdater", e, doneFuture);
+            }
+        }
+    }
+
+    /**
+     * StatusUpdater reads ProcessWorker's stdout line by line. If the line is a JSON object, StatusUpdater updates
+     * the status with the JSON object.
+     */
+    public class StatusUpdater implements Runnable {
+        @Override
+        public void run() {
+            try (BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8))) {
+                String line;
+                while ((line = br.readLine()) != null) {
+                    try {
+                        JsonNode resp = JsonUtil.JSON_SERDE.readTree(line);
+                        if (resp.has("status")) {
+                            status.update(resp.get("status"));
+                        }
+                        if (resp.has("log")) {
+                            log.info("{}: (stdout of the process): {}", id, resp.get("log").toString());
+                        }
+                        if (resp.has("error")) {
+                            JsonNode errNode = resp.get("error");
```

Comment:
I think we should have a `log.info()` here to show the error. I'm afraid users might miss some error statuses once they get overridden, so it's good to have it persisted somewhere

Reply:
Added log.error() for error messages, and log.info() for status updates and log messages.

Reply:
An error is intended to shut down the worker.  So we should call `doneFuture#complete` here.

Reply:
I think it is better to wait for the external process to exit, because out current execution model is that spec workers always finish after one task. So, here, we should expect the external process will exit after finishing the task either successful or with error.